Implement Listeners in CSE and Greedy Pattern Rewriter

Mogball opened the pull request:
In this PR, I've made a demo functional pattern implementation in C++ used it in parts of the LinalgTransform interpreter. Notably, the use of attribute filters and `findMarkedOps` is no longer needed. Although, these specific cases were uninteresting: since the `execute*` functions were passed the Operation handle directly... why not just call the pattern on the handle?

Ideally, mapping `Value` to operations would not be needed at all and the operation handles can be passed around directly. However, since not all of the interpreter transformations are targetable (yet) and I am unsure as to what certain transformations should be returning, I left them as is. If these issues are resolved, the interpreter will disappear entirely and `linalg_transform` operations will just be a description on how to construct and chain the functional patterns.

A few other things to note: I added a small scoping mechanism to restrict the application of patterns, and I reimplemented the tracking greedy rewriting using a listener. I also implemented a rollback mechanism for applying chains of patterns, but it's not used anywhere. It's very inefficient (maintains copies of the IR and associates them) because the dialect conversion system is unwieldy.

nicolasvasilache commented on include/Dialects/LinalgTransform/LinalgTransformOps.td:
```diff
@@ -33,6 +34,34 @@ class Linalg_Transform_Operation<string name, list<OpTrait> props = []>
 
 //===----------------------------------------------------------------------===//
 
+def ScopeOp : Linalg_Transform_Operation<"scope",
```
We were discussing doing exactly this but on FuncOps, with the extra benefit of being able to "name" things.
So I started slicing out functionality (e.g. https://reviews.llvm.org/D116591).

Looking forward, do you see value in:
1. keeping this op without a Symbol interface ?
2. having a separate op than just FuncOp ?

Mogball replied:
This operation is meant to be short-lived. I.e.

```
ScopeOp scope = wrapInScope(myOp);
/// do a bunch of transformations in the scope
unwrapScope(scope);
```

In which case it would not need a symbol to look up later. If it needs to exist longer than a few transformations, then you could pass the handle around instead of passing the name around and looking it up.

Mogball commented on lib/Dialects/LinalgTransform/Transforms/ScopedTransform.cpp:
```diff
@@ -0,0 +1,128 @@
+  BlockAndValueMapping bv;
+  for (auto arg_operand :
+       llvm::zip(body.addArguments(op->getOperandTypes()), op->getOperands()))
+    bv.map(std::get<1>(arg_operand), std::get<0>(arg_operand));
```
Right, thanks!

nicolasvasilache commented on include/Dialects/LinalgTransform/FunctionalTransforms.h:
```diff
@@ -0,0 +1,77 @@
+namespace mlir {
+namespace linalg {
+
+class FunctionalPaddingPattern {
```
I did some refactoring upstream and you should soon be able to just:
```
FailureOr<LinalgOp> applyPaddingPattern(
    LinalgOp op, PatternRewriter&rewriter, const LinalgPaddingOptions &options) {
  LinalgPaddingPattern pattern(op->getContext(), options);
  return pattern.returningMatchAndRewrite(op, rewriter);
}
```
everywhere here and below.

nicolasvasilache replied:
The default empty filter will be passed and we wouldn't need to duplicate all those lines.
Do you have further extensions planned that wouldn't work with this?

nicolasvasilache commented on include/Dialects/LinalgTransform/FunctionalTransforms.h:
```diff
@@ -0,0 +1,77 @@
+  LinalgTilingOptions options;
+};
+
+class FunctionalGenericTilingPattern : FunctionalBaseTilingPattern {
```
There will soon be a single LinalgTilingPattern after https://reviews.llvm.org/D116733 lands.

nicolasvasilache commented on lib/Dialects/LinalgTransform/Transforms/FunctionalTransforms.cpp:
```diff
@@ -0,0 +1,222 @@
+}
+
+/// Peel loops after tiling.
+static void peelLoops(RewriterBase &rewriter, TiledLinalgOp &res,
```
Exposing this function in core.
OTOH if you just wrap the unique LinalgTilingPattern we should be in a better place.

nicolasvasilache commented on include/Transforms/Functional.h:
```diff
@@ -0,0 +1,364 @@
+/// Given a scope, apply a pattern until the first successful match and return
+/// the result.
+template <typename PatternT, typename... Args>
+auto applyOnce(Operation *scope, PatternT &&pattern, Args &&...args) {
```
you could have the scopeop directly and simplify a bit

Mogball replied:
These functions are meant to work in general, and not specifically with `ScopeOp` (although, yes, the naming is confusing...). I intend to keep the functionality separate as long as `ScopeOp` is tied to `linalg_transform` (it doesn't have to; I might make it more first-class).

nicolasvasilache commented on include/Transforms/Functional.h:
```diff
@@ -0,0 +1,364 @@
+  typename Traits::result_t result = failure();
+  scope->walk([pattern = std::forward<PatternT>(pattern), &result, &rewriter,
+               &args...](OpT op) {
+    rewriter.setInsertionPoint(op);
```
changing insertion points without a previous OpBuilder::InsertionGuard is a red flag to me, can you make sure we RAII it properly here and everywhere?

Mogball replied:
It is customary to set the insertion point of the pattern rewriter to the root operation before calling into a pattern. An insertion guard is not needed because it would just reset the insertion point to null (and the rewriter is not used afterwards).

See:
https://github.com/llvm/llvm-project/blob/ab7618914dec6384f24b30ad2d45b82a51647a33/mlir/lib/Rewrite/PatternApplicator.cpp#L188

nicolasvasilache commented on include/Transforms/Functional.h:
```diff
@@ -0,0 +1,364 @@
+  static type unpack(type &&value) { return value; }
+};
+template <typename NestedType>
+struct UnpackFailureOr<FailureOr<NestedType>> {
```
Do we need this rvalue dark-magic?
I.e. is it purely for optimization?
If so is it maybe premature?

Mogball replied:
It's more for correctness. Many types aren't copy-constructable but are move-constructable. I.e., what happens if a pattern decides to return a `FailureOr<unique_ptr<SomeType>>`?

nicolasvasilache commented on include/Transforms/Functional.h:
```diff
@@ -0,0 +1,364 @@
+}
+
+/// Find the nearest proper ancestor that is isolated from above.
+Operation *findNearestScope(Operation *op);
```
just getparentoftype since you went throught the trouble of creating an op?

Mogball replied:
`ScopeOp` is specific to `linalg_transform`. These utilities work in general.

nicolasvasilache commented on include/Transforms/Functional.h:
```diff
@@ -0,0 +1,364 @@
+
+namespace detail {
+template <typename OpT>
+struct IsaOr {
```
I do not understand what this does :)

Mogball replied:
Let me document this in a little bit more detail...

nicolasvasilache reviewed (commented):
Could you start splitting a bit?

The Listener + rewriter + CSE could go in first with some test taken from core that would call this directly?

Then once some of the functional stuff is rebased we should have a much more manageable diff?

Thanks for pushing on this!

Mogball commented:
Yeah I can split things up. Thanks for the review!

Mogball commented on include/Transforms/Functional.h:
```diff
@@ -0,0 +1,364 @@
+} // end namespace detail
+
+template <typename SequenceT, typename... Args>
+auto applySequenceOnce(Operation *scope, SequenceT &&sequence, Args &&...args) {
```
Yeah. This is not efficient.

Mogball commented:
I've split out the changes so that this PR is just the listeners. I've added two tests from MLIR core for the canonicalizer and CSE passes with listeners.

ftynse commented on include/Transforms/ListenerCSE.h:
```diff
@@ -0,0 +1,23 @@
+#define LLVM_IREE_SANDBOX_TRANSFORMS_LISTENERCSE_H
+
+#include "Transforms/Listener.h"
+#include "mlir/IR/Operation.h"
```
Nit: could we forward-declare Operation as well? We are only using it as a pointer in signature.

ftynse commented on lib/Transforms/ListenerCSE.cpp:
```diff
@@ -0,0 +1,279 @@
+
+using namespace mlir;
+
+namespace {
```
Could you please annotate which parts of this are taken from the "main" CSE verbatim and which parts are new? This will help keeping code in sync and eventually merging it.

ftynse commented on lib/Transforms/ListenerGreedyPatternRewriteDriver.cpp:
```diff
@@ -0,0 +1,391 @@
+namespace {
+/// This is a worklist-driven driver for the PatternMatcher, which repeatedly
+/// applies the locally optimal patterns in a roughly "bottom up" way.
+class GreedyPatternRewriteDriver : public RewriteListener {
```
Ditto.

ftynse commented on test/lib/Transforms/TestListenerPasses.cpp:
```diff
@@ -0,0 +1,94 @@
+  }
+};
+
+struct TestCanonicalizePass : public PassWrapper<TestCanonicalizePass, Pass> {
```
Nit: I wouldn't call it "test canonicalize" to (a) avoid confusion with the "main" canonicalize and (b) make it clear that it is testing the rewriter infra rather than canonicalization patterns.

nicolasvasilache commented on lib/Transforms/ListenerCSE.cpp:
```diff
@@ -0,0 +1,310 @@
+using namespace mlir;
+
+//===----------------------------------------------------------------------===//
+// BEGIN copied from mlir/lib/Transforms/CSE.cpp
```
Much better to help review, thanks!

nicolasvasilache commented on test/Transforms/test-listener-canonicalize.mlir:
```diff
@@ -0,0 +1,98 @@
+// RUN: mlir-proto-opt %s -allow-unregistered-dialect -test-listener-canonicalize --split-input-file | FileCheck %s
+
```
Same thing here re copied from ?
Maybe just a top of the file comment that this is fully copied from core and to not modify?

nicolasvasilache commented on test/Transforms/test-listener-cse.mlir:
```diff
@@ -0,0 +1,246 @@
+// RUN: mlir-proto-opt -allow-unregistered-dialect %s -pass-pipeline='builtin.func(test-listener-cse)' | FileCheck %s
+
```
Same thing here re copied from ?
Maybe just a top of the file comment that this is fully copied from core and to not modify?

nicolasvasilache reviewed (approved):
Very nice!

Please merge at your convenience.
